source/extensions/clusters/aggregate_retry/cluster.cc

```diff
@@ -0,0 +1,227 @@
+
+#include "source/common/common/assert.h"
+#include "source/common/common/macros.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Clusters {
+namespace AggregateRetry {
+
+Cluster::Cluster(const envoy::config::cluster::v3::Cluster& cluster,
+                 const envoy::extensions::clusters::aggregate_retry::v3::ClusterConfig& config,
+                 Upstream::ClusterFactoryContext& context, absl::Status& creation_status)
+    : Upstream::ClusterImplBase(cluster, context, creation_status),
+      cluster_manager_(context.serverFactoryContext().clusterManager()),
+      clusters_(std::make_shared<ClusterSet>(config.clusters().begin(), config.clusters().end())),
+      retry_overflow_behavior_(config.retry_overflow_behavior()) {}
+
+AggregateRetryClusterLoadBalancer::AggregateRetryClusterLoadBalancer(
+    const Upstream::ClusterInfoConstSharedPtr& parent_info,
+    Upstream::ClusterManager& cluster_manager, const ClusterSetConstSharedPtr& clusters,
+    envoy::extensions::clusters::aggregate_retry::v3::ClusterConfig::RetryOverflowBehavior
+        retry_overflow_behavior)
+    : parent_info_(parent_info), cluster_manager_(cluster_manager), clusters_(clusters),
+      retry_overflow_behavior_(retry_overflow_behavior) {
+  for (const auto& cluster : *clusters_) {
+    auto tlc = cluster_manager_.getThreadLocalCluster(cluster);
+    // It is possible when initializing the cluster, the included cluster doesn't exist. e.g., the
+    // cluster could be added dynamically by xDS.
+    if (tlc == nullptr) {
+      continue;
+    }
+
+    // Add callback for clusters initialized before aggregate retry cluster.
+    addMemberUpdateCallbackForCluster(*tlc);
+  }
+  refresh();
+  handle_ = cluster_manager_.addThreadLocalClusterUpdateCallbacks(*this);
+}
+
+void AggregateRetryClusterLoadBalancer::addMemberUpdateCallbackForCluster(
+    Upstream::ThreadLocalCluster& thread_local_cluster) {
+  member_update_cbs_[thread_local_cluster.info()->name()] =
+      thread_local_cluster.prioritySet().addMemberUpdateCb(
+          [this, target_cluster_info = thread_local_cluster.info()](const Upstream::HostVector&,
+                                                                    const Upstream::HostVector&) {
+            ENVOY_LOG(debug, "member update for cluster '{}' in aggregate retry cluster '{}'",
+                      target_cluster_info->name(), parent_info_->name());
+            refresh();
+            return absl::OkStatus();
+          });
+}
+
+uint32_t AggregateRetryClusterLoadBalancer::getRetryAttemptCount(
+    Upstream::LoadBalancerContext* context) const {
+  if (context == nullptr) {
+    return 0;
+  }
+
+  // Get retry attempt count from stream info.
+  auto* stream_info = context->requestStreamInfo();
+  if (stream_info != nullptr && stream_info->attemptCount().has_value()) {
+    return stream_info->attemptCount().value();
+  }
+
+  return 0;
+}
+
+size_t
```

Comment:
nit: I suggest instead of a sentinel value to use either StatusOr or optional to indicate terminal state. Out of bound indexes can be error prone if someone does not check this in some future changes.